dagger-core/src/main/java/io/odpf/dagger/core/source/parquet/reader/ParquetReader.java

```diff
@@ -0,0 +1,126 @@
+import org.apache.parquet.example.data.simple.convert.GroupRecordConverter;
+import org.apache.parquet.hadoop.ParquetFileReader;
+import org.apache.parquet.hadoop.util.HadoopInputFile;
+import org.apache.parquet.io.ColumnIOFactory;
+import org.apache.parquet.io.MessageColumnIO;
+import org.apache.parquet.io.RecordReader;
+import org.apache.parquet.schema.MessageType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+
+public class ParquetReader implements FileRecordFormat.Reader<Row> {
+    private final Path hadoopFilePath;
+    private final SimpleGroupDeserializer simpleGroupDeserializer;
+    private long currentRecordIndex;
+    private final ParquetFileReader parquetFileReader;
+    private long rowCount;
+    private boolean isRecordReaderInitialized;
+    private RecordReader<Group> recordReader;
+    private final MessageType schema;
+    private static final Logger LOGGER = LoggerFactory.getLogger(ParquetReader.class.getName());
+
+    private ParquetReader(Path hadoopFilePath, SimpleGroupDeserializer simpleGroupDeserializer, ParquetFileReader parquetFileReader) throws IOException {
+        this.hadoopFilePath = hadoopFilePath;
+        this.simpleGroupDeserializer = simpleGroupDeserializer;
+        this.parquetFileReader = parquetFileReader;
+        this.schema = this.parquetFileReader.getFileMetaData().getSchema();
+        this.isRecordReaderInitialized = false;
+    }
+
+    private void changeReaderPosition(PageReadStore pages) {
+        rowCount = pages.getRowCount();
+        currentRecordIndex = 0;
+        MessageColumnIO columnIO = new ColumnIOFactory().getColumnIO(schema);
+        recordReader = columnIO.getRecordReader(pages, new GroupRecordConverter(schema));
+    }
+
+    private Row initializeRecordReader() throws IOException {
+        PageReadStore page = parquetFileReader.readNextRowGroup();
+        if (page == null) {
+            String logMessage = String.format("No readable data found in Parquet file %s", hadoopFilePath.getName());
+            LOGGER.info(logMessage);
+            return null;
+        }
+        changeReaderPosition(page);
+        this.isRecordReaderInitialized = true;
+        String logMessage = String.format("Successfully created the ParquetFileReader and RecordReader for file %s", hadoopFilePath.getName());
+        LOGGER.info(logMessage);
+        return readAndDeserialize();
+    }
+
+    private Row readRecordsFromPage() throws IOException {
+        if (currentRecordIndex >= rowCount) {
+            PageReadStore page = parquetFileReader.readNextRowGroup();
+            if (page == null) {
+                return null;
+            }
+            changeReaderPosition(page);
+        }
+        return readAndDeserialize();
+    }
+
+    @Nullable
+    @Override
+    public Row read() throws IOException {
+        if (!isRecordReaderInitialized) {
+            return initializeRecordReader();
```

Comment:
at present, initialize is doing multiple things. Initializing as well as reading. Can we break it down and rather just use readRecordsFromPage after initialization is done? 
so that way, you only initialize when it is not initialized and the rest of the flow is the same irrespective of flow (first time or later).